### PR TITLE
PXB-2977 Validate decompress paths (qpress, zstd)

### PR DESCRIPTION
**Pull request by rahulmalik87**
https://jira.percona.com/browse/PXB-2977

Problem:
While decompressing a backup, we pass filepath and dest_filepath to system call without any validation.

Fix:
Pass file name as enviorment variable so shell does not extract that

### CODE REVIEW

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: setenv can fail, do we need to check return code?

Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: If we are setting environment variables, any chance that two --decompress on the same shell will conflict with each other?

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: same here, we should check return code.

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: shall we use unsetenv once we are done so we don't leave any env var set behind?

**Review by altmannmarcelo — CHANGES_REQUESTED** [review]
Hi @rahulmalik87 Interesting way of fixing the issue.

please address review comments and post jenkins results

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: for each thread N I am using xbinputfileN. so not possible.

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: fixed

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: fixed

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: I did not mean threads, I meant two xtrabackups

xtrabackup --decompress --target-dir=dir1 &
xtrabackup --decompress --target-dir=dir2 &

Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: this Environment variables are set only for the process.  A processes can not access the environment variables set for other processes

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: No need to store it on a variable (see comment about `msg()`)

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: Minor fixes here:
 - Can we use msg_ts
 - Can we follow the standard of other errors that print the thread nr issuing the msg
 - Can we also set which file we were processing (`dest_filepath`)
 - Variable `env_error` will always be -1 in case of error.  `setenv` also sets [errno](https://man7.org/linux/man-pages/man3/errno.3.html) . A better user experience in case of error will be to return the actual error, rather than always -1.
 - With above, we don't need to store the return of `setenv`, just test it within the if clause.

**Review by altmannmarcelo — CHANGES_REQUESTED** [review]
some improvements on error message.

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/backup_copy.cc:2181]: From our discussion on slack, escape_string_for_mysql() didn't work according to our expectations. Can we fix this instead?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/test/t/xb_decompress_file_validation.sh
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/t/xb_decompress_file_validation.sh]: remove all these lines

File: storage/innobase/xtrabackup/test/t/xb_decompress_file_validation.sh
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/t/xb_decompress_file_validation.sh:8]: I assume the fix path is common to other compression libraries, Can you please check with zstd?

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: it would be nice to use capitalnames for environmental variables.

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: Please add a comment how the final command line will look like

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (satya-bodapati) [reply storage/innobase/xtrabackup/src/backup_copy.cc:2181]: Lets take your approach.

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: I choose small letter environmental variables because OS   Uses capital letter and it doesn't interfere with them

File: storage/innobase/xtrabackup/test/t/xb_decompress_file_validation.sh
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/t/xb_decompress_file_validation.sh:8]: yes, for 8.0. the test is with zstd . For 2.4 It is with qpress

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: added

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (satya-bodapati) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: better to use capital letters, it wouldn't cause conflict. Please PXBOUTFILE and PXBINFILE

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: Not here rahul and not like this.

Comment (satya-bodapati) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: Just before system() call.  cat $XBINFILE|xbcrtyp --decrpyt|qpress -dio > $XBOUTFILE

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
Almost there.

Please get the Jenkins link

File: storage/innobase/xtrabackup/src/backup_copy.cc
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: typo: xbcrypt

**Review by satya-bodapati — APPROVED** [review]
Please wait for @altmannmarcelo's approval before merge
I assume the 8.0 patch would change in test case (additional zstd, scenario?)

**Review by altmannmarcelo — APPROVED** [review]
LGTM. Since we modified a few parts of the original patch. Can you please submit a new Jenkins ?

**Pull request closed without merging**
